code/game/machinery/doors/firedoor.dm

```diff
@@ -203,30 +199,90 @@
+	SIGNAL_HANDLER
+	var/datum/merger/temp_group = source
+	var/datum/merger/merge_group = GetMergeGroup(merger_id, merger_typecache)
+	if(temp_group.origin != src)
+		return
+	var/list/shared_problems = list() // We only want to do this once, this is a nice way of pulling that off
+	for(var/obj/machinery/door/firedoor/firelock as anything in merge_group.members)
+		firelock.issue_turfs = shared_problems
+
+/obj/machinery/door/firedoor/proc/register_adjacent_turfs(atom/loc)
+	for(var/dir in GLOB.cardinals)
+		var/turf/checked_turf = get_step(get_turf(loc),dir)
+		process_results(checked_turf)
+		if(checked_turf)
+			RegisterSignal(checked_turf, COMSIG_TURF_EXPOSE, .proc/process_results)
+			RegisterSignal(checked_turf, COMSIG_TURF_CALCULATED_ADJACENT_ATMOS, .proc/process_results)
+
+
+/obj/machinery/door/firedoor/proc/unregister_adjacent_turfs(atom/loc)
+	for(var/dir in GLOB.cardinals)
+		var/turf/checked_turf = get_step(get_turf(loc),dir)
+		process_results(checked_turf)
+		if(checked_turf)
+			UnregisterSignal(checked_turf, COMSIG_TURF_EXPOSE)
+			UnregisterSignal(checked_turf, COMSIG_TURF_CALCULATED_ADJACENT_ATMOS)
+
+/obj/machinery/door/firedoor/proc/check_atmos(turf/checked_turf)
+	var/datum/gas_mixture/environment = checked_turf.return_air()
+	var/obj/machinery/airalarm/air_alarm = LAZYACCESS(my_area.airalarms,1)
+	
+	if(air_alarm)
+		var/datum/tlv/selected_temperature = air_alarm.TLV["temperature"]
+		if(environment?.temperature >= selected_temperature?.hazard_max)
+			return FIRELOCK_ALARM_TYPE_HOT
+		if(environment?.temperature <= selected_temperature?.hazard_min)
+			return FIRELOCK_ALARM_TYPE_COLD
+	else
+		if(environment?.temperature >= FIRE_MINIMUM_TEMPERATURE_TO_EXIST)
+			return FIRELOCK_ALARM_TYPE_HOT
+		if(environment?.temperature <= BODYTEMP_COLD_DAMAGE_LIMIT)
+			return FIRELOCK_ALARM_TYPE_COLD
+	return
+
+/obj/machinery/door/firedoor/proc/process_results(datum/source)
+	SIGNAL_HANDLER
+
 	if(!COOLDOWN_FINISHED(src, detect_cooldown))
 		return
-	if(alarm_type)
+	if(alarm_type == FIRELOCK_ALARM_TYPE_GENERIC)
 		return
+
 	for(var/area/place in affecting_areas)
 		if(!place.fire_detect) //if any area is set to disable detection
 			return
 
-	var/turf/my_turf = source
-	if(!my_turf)
-		my_turf = get_turf(src)
-
-	var/datum/gas_mixture/environment = my_turf.return_air()
-	var/result
-
-	if(environment?.temperature >= FIRE_MINIMUM_TEMPERATURE_TO_EXIST)
-		result = FIRELOCK_ALARM_TYPE_HOT
-	if(environment?.temperature <= BODYTEMP_COLD_DAMAGE_LIMIT)
-		result = FIRELOCK_ALARM_TYPE_COLD
-	if(!result)
+	var/turf/checked_turf = source
+	var/result = check_atmos(checked_turf)
+	
+	if(!issue_turfs?.len && alarm_type)
+		start_deactivation_process()
 		return
-
-	start_activation_process(result)
+	else if(issue_turfs?.len && !alarm_type)
+		result = check_atmos(issue_turfs[1])
+		if(result)
+			start_activation_process(result)
+			return
```

Comment:
what's going on here?

Reply:
if we dont give it a code, it breaks other firelocks, so we use the code the first issue we found has